### PR TITLE
KAFKA-7502: Cleanup KTable materialization logic in a single place (filter)

### PR DESCRIPTION
**Pull request by dongjinleekr**
This PR is a follow-up of #6174, which handles `doFilter` method.

cc/ @guozhangwang @bbejeck

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bbejeck** [comment]
Java 8 passed, Java 11 failed.  Results already cleaned up.

retest this please

**Comment by bbejeck** [comment]
ping @guozhangwang @mjsax @vvcephei @ableegoldman for reviews

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 1]
```diff
@@ -114,17 +114,36 @@ public String queryableStoreName() {
...
-        // we actually do not need to generate store names at all since if it is not specified, we will not
-        // materialize the store; but we still need to burn one index BEFORE generating the processor to keep compatibility.
-        if (materializedInternal != null && materializedInternal.storeName() == null) {
-            builder.newStoreName(FILTER_NAME);
+        final Serde<K> keySerde;
+        final Serde<V> valueSerde;
+        final String queryableStoreName;
+        final StoreBuilder<KeyValueStore<K, V>> storeBuilder;
+
+        // we can inherit parent key and value serde if user do not provide specific overrides, more specifically:
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: nit: move the comments regarding serde inheritance directly above the lines where each respective check is done

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java [hunk 2]
```diff
@@ -17,36 +17,32 @@
...
-    private final MaterializedInternal<K, V, S> materializedInternal;
     private final ProcessorParameters<K, V> processorParameters;
     private final String[] storeNames;
+    private final StoreBuilder<KeyValueStore<K, V>> storeBuilder;
 
     public TableProcessorNode(final String nodeName,
                               final ProcessorParameters<K, V> processorParameters,
-                              final MaterializedInternal<K, V, S> materializedInternal,
-                              final String[] storeNames) {
+                              final String[] storeNames,
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java]: why the order change of the parameters?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java [hunk 3]
```diff
@@ -62,15 +58,9 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
...
-        final boolean shouldMaterialize = materializedInternal != null && materializedInternal.queryableStoreName() != null;
-        if (shouldMaterialize) {
-            // TODO: we are enforcing this as a keyvalue store, but it should go beyond any type of stores
-            topologyBuilder.addStateStore(
-                new KeyValueStoreMaterializer<>(
-                    (MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>>) materializedInternal
-                ).materialize(),
-                processorName);
+        // TODO: we are enforcing this as a keyvalue store, but it should go beyond any type of stores
+        if (this.storeBuilder != null) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java]: nit: remove `this` here and line below

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java [hunk 4]
```diff
@@ -17,36 +17,32 @@
...
 
-public class TableProcessorNode<K, V, S extends StateStore> extends StreamsGraphNode {
+public class TableProcessorNode<K, V> extends StreamsGraphNode {
 
-    private final MaterializedInternal<K, V, S> materializedInternal;
     private final ProcessorParameters<K, V> processorParameters;
     private final String[] storeNames;
+    private final StoreBuilder<KeyValueStore<K, V>> storeBuilder;
 
     public TableProcessorNode(final String nodeName,
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java:32]: Although the constructor was pre-existing, I'm thinking we could clean things up a little bit by adding a constructor
```java
 public TableProcessorNode(final String nodeName,
        final ProcessorParameters<K, V> processorParameters,
        final StoreBuilder<KeyValueStore<K, V>> storeBuilder) {
        
        this(nodeName, processorParameters, null, storeBuilder);
    }
```
Then it's more clear in the code when we call 
```java
final StreamsGraphNode tableNode = new TableProcessorNode<>(
            name,
            processorParameters,
            storeBuilder
        );
```
And we can leave the existing constructor using all 4 parameters alone.

**Review by bbejeck — APPROVED** [review]
Thanks @dongjinleekr, I've made a pass over overall this looks good to me.  I just have a couple of minor comments.

**Review by bbejeck — COMMENTED** [review]
Actually meant to hit comment first.

Thanks @dongjinleekr, I've made a pass over overall this looks good to me. I just have a couple of minor comments.

**Review by bbejeck — COMMENTED** [review]
Thanks for the updates @dongjinleekr looks good to me.

However, the build failure is related.
Take a look at line 314 in `KTableImpl` the `TableProcessorNode` constructor parameters need adjusting.

**Comment by dongjinleekr** [comment]
@bbejeck I'm sorry, I omitted a line from commit; here is the fix with rebasing onto the latest trunk.
Screenshot from 2019-03-25 16-00-56

**Review by bbejeck — APPROVED** [review]
Thanks for the updates @dongjinleekr, LGTM

**Comment by bbejeck** [comment]
call for second review @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 5]
```diff
@@ -309,8 +325,8 @@ public String queryableStoreName() {
         // we preserve the value following the order of 1) materialized, 2) null
         return new KTableImpl<>(
             name,
-            materialized != null && materialized.keySerde() != null ? materialized.keySerde() : keySerde,
-            materialized != null ? materialized.valueSerde() : null,
+            materializedInternal != null && materializedInternal.keySerde() != null ? materializedInternal.keySerde() : keySerde,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:328]: nit: the coding style in `filter` and `mapValues` are slightly different here: can we just make it consistent to pre-create the serdes beforehand as we did in `doFilter`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java [hunk 6]
```diff
@@ -17,39 +17,40 @@
...
     }
 
     @Override
     public String toString() {
         return "TableProcessorNode{" +
-               "materializedInternal=" + materializedInternal +
-               ", processorParameters=" + processorParameters +
-               ", storeNames=" + Arrays.toString(storeNames) +
-               "} " + super.toString();
+            ", processorParameters=" + processorParameters +
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java:51]: Print `storeBuilder.name()` here as well?

**Review by guozhangwang — COMMENTED** [review]
Left two comments, otherwise LGTM!

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 5 above)
Comment (dongjinleekr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:328]: Sure. It is already done in the other branch which handles `transformValues` method, not here. :smiley:

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java (hunk 6 above)
Comment (dongjinleekr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java:51]: No problem. :)

**Comment by guozhangwang** [comment]
Jenkins failure: https://issues.apache.org/jira/browse/KAFKA-7965

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
Thanks @dongjinleekr !

**Pull request merged**